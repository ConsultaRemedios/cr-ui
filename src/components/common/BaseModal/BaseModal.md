A raw Modal component. This component does not control the style inside the modal. It just
renders a Modal with a close button (an icon).

### Usage normal
```js
new Vue({
  data(){
    return { showModal: false };
  },
  methods: {
    openModal() {
      this.showModal = true;
    },

    closeModal() {
      this.showModal = false;
    },
  },
  template: `
    <div>
      <BaseModal v-if="showModal" @close="closeModal">
        <div :style="{ padding: '20px' }">
          <h2>Base Modal</h2>
          <p>This is a raw base modal. It's uggly, I know.</p>
        </div>
      </BaseModal>

      <BaseButton @click="openModal">Open Modal</BaseButton>
    </div>
  `
})
```
### Usage footer modal
```js
new Vue({
  data(){
    return { showModal: false };
  },
  methods: {
    openModal() {
      this.showModal = true;
    },

    closeModal() {
      this.showModal = false;
    },
  },
  template: `
    <div>
      <BaseModal v-if="showModal" @close="closeModal" type="footer">
        <div :style="{ padding: '20px' }">
          <h2>Base Modal</h2>
          <p>This is a raw base modal. It's uggly, I know.</p>
        </div>
      </BaseModal>

      <BaseButton @click="openModal">Open Modal</BaseButton>
    </div>
  `
})
```

### Dismissible
You can set dismissible prop as `true` if you don't want a close button.
Also, The BaseModal component will not trigger `close` event when clicking on overlay.

```js
new Vue({
  data(){
    return { showModal: false };
  },
  methods: {
    openModal() {
      this.showModal = true;
    },

    closeModal() {
      this.showModal = false;
    },
  },
  template: `
    <div>
      <BaseModal v-if="showModal" :dismissible="false">
        <div :style="{ padding: '20px' }">
          <h2>Base Modal</h2>
          <p>This is a raw base modal. It's uggly, I know.</p>

          <BaseButton @click="closeModal">Close Modal</BaseButton>
        </div>
      </BaseModal>

      <BaseButton @click="openModal">Open Modal</BaseButton>
    </div>
  `
})
```
