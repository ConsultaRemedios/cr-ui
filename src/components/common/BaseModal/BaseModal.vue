<template>
  <transition
    :enter-class="$style.enterTransition"
    :leave-active-class="$style.leaveActiveTransition"
  >
    <div :class="[$style.overlay, 'cr-base-modal__overlay', customCssClass]" @click="onClickOverlay" data-modal-overlay>
      <div :class="[$style.container, 'cr-base-modal__container']" data-modal-container>
        <div :class="[$style.content, 'cr-base-modal__content']" data-modal-content>
          <button v-if="dismissible" data-modal-close
            :class="[$style.closeButton, 'cr-base-modall__close']"
            @click="$emit('close')"
          >
            <span :class="$style.closeButtonLabel">Fechar</span>
          </button>

          <slot></slot>
        </div>
      </div>
    </div>
  </transition>
</template>

<script>
  import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';

  export default {
    name: 'BaseModal',
    props: {
      dismissible: {
        type: Boolean,
        default: true,
      },

      customCssClass: {
        type: Array,
        default: () => [],
      },
    },

    methods: {
      onClickOverlay(e) {
        const isOverlayTheTarget = e.target.querySelector('[data-modal-content]');

        if (isOverlayTheTarget && this.dismissible) {
          this.$emit('close');
        }
      },
    },

    mounted() {
      disableBodyScroll(this.$el);
    },

    destroyed() {
      enableBodyScroll(this.$el);
    },
  };
</script>

<style module>
  .overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.6);
    z-index: 9999;
    overflow: auto;

    transition: opacity 0.2s ease-out;
    opacity: 1;
  }

  .container {
    min-width: 100%;
    min-height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 20px;
  }

  .content {
    background: #FFF;
    position: relative;
    border-radius: 4px;
    transition: transform 0.2s ease-out;
    transform: scale(1);
  }

  .closeButton {
    position: absolute;
    top: 15px;
    right: 15px;
    cursor: pointer;
    padding: 5px;
    background: none;
    border: 0;
  }

  .closeButton::after {
    content: '';
    display: block;
    width: 14px;
    height: 14px;
    background: url('./../../../assets/close-icon.svg');
  }

  .closeButtonLabel {
    display: block;
    font-size: 0;
    height: 1px;
    overflow: hidden;
  }

  .enterTransition,
  .leaveActiveTransition {
    opacity: 0;
  }

  .enterTransition .content,
  .leaveActiveTransition .content {
    transform: scale(1.05);
  }

  @media (min-width: 768px) {
    .closeButton {
      top: 20px;
      right: 20px;
    }
  }
</style>
