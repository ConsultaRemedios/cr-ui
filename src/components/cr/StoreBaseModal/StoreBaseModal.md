It is an inner component currently used by <a href="/#StoreModal">StoreModal</a> and <a href="/#StorePolicyModal">StorePolicyModal</a>.
