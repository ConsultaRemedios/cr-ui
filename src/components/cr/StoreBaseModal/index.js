export { default } from './StoreBaseModal.vue';
