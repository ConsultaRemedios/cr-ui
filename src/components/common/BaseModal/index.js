import BaseModal from './BaseModal.vue';

export default BaseModal;
