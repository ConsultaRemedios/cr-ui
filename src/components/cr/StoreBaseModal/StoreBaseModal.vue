<template>
  <BaseModal @close="$emit('close')">
    <div
      :class="[
        $style.wrapper,
        !isLoading && $style.loadingMaxHeight,
      ]"
      :style="{ maxHeight: `${ui.maxContentHeight}px` }"
    >
      <div>
        <span :class="$style.title">
          {{ title }}
        </span>
        <img
          :class="$style.logo"
          :src="logoPath"
          :alt="name"
          @load="updateMaxContentHeight"
          @error="updateMaxContentHeight"
        >
      </div>

      <div
        :class="[$style.content, {
            [$style.showScroll]: isMobileBreakpoint,
            [$style.contentloader]: isLoading,
          }]"
      >
        <SimpleLoader
          v-if="isLoading"
        />

        <div v-else>
          <slot></slot>
        </div>
      </div>
    </div>
  </BaseModal>
</template>

<script>
  import BaseModal from '../../common/BaseModal';
  import SimpleLoader from '../../common/SimpleLoader';
  import breakpointable from '../../../mixins/breakpointable';

  const modalExternalPadding = 40;

  export default {
    name: 'StoreBaseModal',

    components: {
      BaseModal,
      SimpleLoader,
    },

    mixins: [breakpointable],

    props: {
      title: {
        type: String,
        default: '',
      },

      name: {
        type: String,
        default: '',
      },

      logoPath: {
        type: String,
        default: '',
      },

      isLoading: {
        type: Boolean,
        default: false,
      },
    },

    data() {
      return {
        ui: {
          maxContentHeight: 0,
        },
      };
    },

    methods: {
      updateMaxContentHeight() {
        this.ui.maxContentHeight = window.innerHeight - (modalExternalPadding * 2);
      },
    },

    mounted() {
      this.updateMaxContentHeight();

      window.addEventListener('resize', this.updateMaxContentHeight);
    },

    destroyed() {
      window.removeEventListener('resize', this.updateMaxContentHeight);
    },
  };
</script>

<style module>
  .loadingMaxHeight {
    max-height: 116px;
  }

  .wrapper {
    display: flex;
    flex-direction: column;
    max-height: 2000px;
    transition: max-height 0.15s ease-out;
    padding: 20px;
  }

  .title {
    display: block;
    min-width: 200px;
    font-size: 16px;
    font-weight: 600;
    margin-bottom: 10px;
    color: #333333;
  }

  .logo {
    margin-bottom: 15px;
    max-width: 130px;
    vertical-align: middle;
  }

  .contentloader {
    display: flex;
    align-items: center;
    justify-content: center;
    min-width: 260px;
    min-height: 260px;
  }

  .content {
    overflow: auto;
    overflow-wrap: break-word;
    word-wrap: break-word;
    word-break: break-word;
  }

  .showScroll {
    overflow: scroll;
  }

  .showScroll::-webkit-scrollbar {
    -webkit-appearance: none;
  }

  .showScroll::-webkit-scrollbar:vertical {
    width: 5px;
  }

  .showScroll::-webkit-scrollbar-thumb {
    background-color: rgba(0, 0, 0, .5);
    border-radius: 10px;
    border: 2px solid transparent;
  }

  .showScroll::-webkit-scrollbar-track {
    border-radius: 10px;
    background-color: transparent;
  }

  @media (min-width: 768px) {
    .wrapper {
      width: 585px;
      padding: 40px;
    }
  }
</style>
