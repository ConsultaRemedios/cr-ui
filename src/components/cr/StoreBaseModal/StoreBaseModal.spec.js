import { mount, shallowMount } from '@vue/test-utils';
import BaseModal from '../../common/BaseModal';
import StoreBaseModal from './StoreBaseModal.vue';

const defaultProps = {
  information: '<p>Some content</p>',
};

const shallowStoreBaseModal = (data = {}, methods = {}, computed = {}) => {
  return shallowMount(StoreBaseModal, {
    propsData: { ...defaultProps, ...data },
    methods,
    computed,
  });
};

const mountStoreBaseModal = () => {
  return mount(StoreBaseModal, {
    propsData: { ...defaultProps },
  });
};

describe('StoreBaseModal component', () => {
  beforeAll(() => {
    global.window.innerHeight = 1000;
  });

  describe('Snapshot', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallowStoreBaseModal();
    });

    it('matches when is on mobile', () => {
      const wrapper = shallowStoreBaseModal(
        {},
        {},
        { isMobileBreakpoint() { return true; } }
      );
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches when is on desktop', () => {
      const wrapper = shallowStoreBaseModal(
        {},
        {},
        { isMobileBreakpoint() { return false; } }
      );
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches when is loading', () => {
      wrapper.setProps({ isLoading: true });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches when is not loading', () => {
      wrapper.setProps({ isLoading: false });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Lifecycle', () => {
    describe('#mounted', () => {
      const updateMaxContentHeight = jest.fn();
      let wrapper;

      beforeAll(() => {
        window.addEventListener = jest.fn();
        wrapper = shallowStoreBaseModal({}, { updateMaxContentHeight });
      });

      it('calls #updateMaxContentHeight', () => {
        expect(updateMaxContentHeight).toHaveBeenCalled();
      });

      it('adds event listener "resize"', () => {
        expect(window.addEventListener).toHaveBeenCalledWith(
          'resize',
          wrapper.vm.updateMaxContentHeight,
        );
      });
    });

    describe('#destroyed', () => {
      let wrapper;

      beforeAll(() => {
        window.removeEventListener = jest.fn();
        wrapper = shallowStoreBaseModal();
        wrapper.destroy()
      });

      it('removes event listener "resize"', () => {
        expect(window.removeEventListener).toHaveBeenCalledWith(
          'resize',
          wrapper.vm.updateMaxContentHeight,
        );
      });
    });
  });

  describe('Event Listeners', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = mountStoreBaseModal();
    });

    describe('When BaseModal emits "close"', () => {
      beforeEach(() => {
        jest.spyOn(wrapper.vm, '$emit');
        wrapper.find(BaseModal).vm.$emit('close');
      });

      it('emits "close"', () => {
        expect(wrapper.vm.$emit).toHaveBeenLastCalledWith('close');
      });
    });
  });

  describe('Methods', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallowStoreBaseModal();
    });

    describe('#updateMaxContentHeight', () => {
      describe('When window.innerHeight is 500', () => {
        beforeEach(() => {
          global.window.innerHeight = 500;
          wrapper.vm.updateMaxContentHeight();
        });

        it('updates "ui.maxContentHeight" to 420', () => {
          expect(wrapper.vm.ui.maxContentHeight).toBe(420);
        });
      });
    });
  });
});
