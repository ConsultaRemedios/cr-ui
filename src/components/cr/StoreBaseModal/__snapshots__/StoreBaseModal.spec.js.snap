// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreBaseModal component Snapshot matches when is loading 1`] = `
<basemodal-stub
  dismissible="true"
  information="<p>Some content</p>"
>
  <div
    class="wrapper"
    style="max-height: 920px;"
  >
    <div>
      <span
        class="title"
      >
        
        
      
      </span>
       
      <img
        alt=""
        class="logo"
        src=""
      />
    </div>
     
    <div
      class="content contentloader"
    >
      <simpleloader-stub />
    </div>
  </div>
</basemodal-stub>
`;

exports[`StoreBaseModal component Snapshot matches when is not loading 1`] = `
<basemodal-stub
  dismissible="true"
  information="<p>Some content</p>"
>
  <div
    class="wrapper loadingMaxHeight"
    style="max-height: 920px;"
  >
    <div>
      <span
        class="title"
      >
        
        
      
      </span>
       
      <img
        alt=""
        class="logo"
        src=""
      />
    </div>
     
    <div
      class="content"
    >
      <div />
    </div>
  </div>
</basemodal-stub>
`;

exports[`StoreBaseModal component Snapshot matches when is on desktop 1`] = `
<basemodal-stub
  dismissible="true"
  information="<p>Some content</p>"
>
  <div
    class="wrapper loadingMaxHeight"
    style="max-height: 920px;"
  >
    <div>
      <span
        class="title"
      >
        
        
      
      </span>
       
      <img
        alt=""
        class="logo"
        src=""
      />
    </div>
     
    <div
      class="content"
    >
      <div />
    </div>
  </div>
</basemodal-stub>
`;

exports[`StoreBaseModal component Snapshot matches when is on mobile 1`] = `
<basemodal-stub
  dismissible="true"
  information="<p>Some content</p>"
>
  <div
    class="wrapper loadingMaxHeight"
    style="max-height: 920px;"
  >
    <div>
      <span
        class="title"
      >
        
        
      
      </span>
       
      <img
        alt=""
        class="logo"
        src=""
      />
    </div>
     
    <div
      class="content showScroll"
    >
      <div />
    </div>
  </div>
</basemodal-stub>
`;
